mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala

```diff
@@ -0,0 +1,466 @@
+        val msg = s"${optimizer.getClass.getName} failed."
+        throw new SparkException(msg)
+      }
+
+      state.x.toArray.clone()
+    }
+
+    if (handlePersistence) instances.unpersist()
+
+    val rawCoefficients = parameters.slice(2, parameters.length)
+    var i = 0
+    while (i < numFeatures) {
+      rawCoefficients(i) *= { if (featuresStd(i) != 0.0) 1.0 / featuresStd(i) else 0.0 }
+      i += 1
+    }
+    val coefficients = Vectors.dense(rawCoefficients)
+    val intercept = parameters(1)
+    val scale = parameters(0)
+    val model = new RobustRegressionModel(uid, coefficients, intercept, scale)
+    copyValues(model.setParent(this))
+  }
+
+  @Since("2.1.0")
+  override def copy(extra: ParamMap): RobustRegression = defaultCopy(extra)
+
+}
+
+/**
+ * Model produced by [[RobustRegression]].
+ */
+@Since("2.1.0")
+class RobustRegressionModel private[ml] (
+     @Since("2.1.0") override val uid: String,
+     @Since("2.1.0") val coefficients: Vector,
+     @Since("2.1.0") val intercept: Double,
+     @Since("2.1.0") val scale: Double)
+  extends RegressionModel[Vector, RobustRegressionModel] with LinearRegressionParams {
+
+  override protected def predict(features: Vector): Double = {
+    dot(features, coefficients) + intercept
+  }
+
+  @Since("2.1.0")
+  override def copy(extra: ParamMap): RobustRegressionModel = {
+    val newModel = copyValues(new RobustRegressionModel(uid, coefficients, intercept, scale), extra)
+    newModel.setParent(parent)
+  }
+}
+
+/**
+ * HuberAggregator computes the gradient and loss for a huber loss function,
+ * as used in robust regression for samples in sparse or dense vector in an online fashion.
+ *
+ * The huber loss function based on:
+ * Art B. Owen (2006), A robust hybrid of lasso and ridge regression.
+ * ([[http://statweb.stanford.edu/~owen/reports/hhu.pdf]])
+ *
+ * Two HuberAggregator can be merged together to have a summary of loss and gradient of
+ * the corresponding joint dataset.
+ *
+ * The huber loss function is given by
+ * {{{
+ *   {min\,} {\sum_{i=1}^n\left(\sigma +
+ *   H_m\left(\frac{X_{i}w - y_{i}}{\sigma}\right)\sigma\right) + \alpha {||w||_2}^2}
+ * }}}
+ * where
+ * {{{
+ *   H_m(z) = \begin{cases}
+ *            z^2, & \text {if } |z| < \epsilon, \\
+ *            2\epsilon|z| - \epsilon^2, & \text{otherwise}
+ *            \end{cases}
+ * }}}
+ *
+ * @param parameters including three part: The scale parameter (sigma), the intercept and
+ *                regression coefficients corresponding to the features.
+ * @param fitIntercept Whether to fit an intercept term.
+ * @param featuresStd The standard deviation values of the features.
+ * @param m The shape parameter to control the amount of robustness.
+ */
+private class HuberAggregator(
+    parameters: Vector,
+    fitIntercept: Boolean,
+    featuresStd: Array[Double],
+    m: Double) extends Serializable {
+
+  private val coefficients: Array[Double] = parameters.toArray.slice(2, parameters.size)
```

Comment:
This aggregator will serialize the `featuresStd` and the `coefficients` between aggregation steps, which is not necessary. You can mark them as `@transient` or simply pass them to the add function as `LogisticRegression` does. You can see https://github.com/apache/spark/pull/14109 for more details.